Regel voor gebruik RFC 9457

sanderke opened the pull request:
De linter bevatte reeds een regel voor het verschaffen van "problem details". Hier wordt het ook als design rule toegevoegd.

fixes #188

sanderke committed:
Add rule

sanderke committed:
Formatting

sanderke committed:
markdownlint

TimvdLippe commented on sections/designRules.md:
```diff
@@ -491,6 +491,38 @@ An API is as good as the accompanying documentation. The documentation has to be
+      </dd>
+      <dt>How to test</dt>
+      <dd>
+         Verify all responses with status code <code>4xx</code> or <code>5xx</code> have <code>Content-Type</code> set to <code>application/problem+json</code> or <code>application/problem+xml</code>.
```
Hier moeten we ook aan toevoegen dat alle fields present zijn, omdat die bij de RFC allemaal optioneel zijn, maar met deze regel allemaal verplicht worden.

sanderke replied:
De velden zijn nu ook hier vermeld.

TimvdLippe commented on sections/designRules.md:
```diff
@@ -491,6 +491,38 @@ An API is as good as the accompanying documentation. The documentation has to be
+   <dl>
+      <dt>Statement</dt>
+      <dd>
+         Error responses with HTTP status codes <code>4xx</code> or <code>5xx</code> MUST use either <code>application/problem+json</code> or <code>application/problem+xml</code> as the <code>Content-Type</code> header, and the response body <strong>MUST</strong> conform to the structure defined in [[rfc9457]].
```
```suggestion
         Error responses with HTTP status codes <code>4xx</code> or <code>5xx</code> MUST use either <code>application/problem+json</code> or <code>application/problem+xml</code> as the <code>Content-Type</code> header, and the response body <strong>MUST</strong> conform to the structure defined in [[rfc9457]] with all fields present.
```

TimvdLippe reviewed (dismissed):
https://github.com/Logius-standaarden/API-Design-Rules/blob/b6f040ac6c3827c4a9f7b9e1e99368b2cf062a09/linter/spectral.yml#L159 moet ook een verwijzing krijgen naar deze nieuwe regel

sanderke committed:
Apply suggestions from code review

Co-authored-by: Tim van der Lippe <TimvdLippe@users.noreply.github.com>

sanderke committed:
Added rule tag

TimvdLippe commented on sections/designRules.md:
```diff
@@ -491,6 +491,37 @@ An API is as good as the accompanying documentation. The documentation has to be
+         <div class="example">
+            The following example shows the head and body of a detailed error response.
+            <pre><code class="http">HTTP/1.1 404 Not Found
+Content-Type: application/problem+json</code><code class="json">
+{
```
```suggestion
Content-Type: application/problem+json</code><code class="json">{
```

Het lijkt er op dat het allemaal achter elkaar moet staan, omdat er bij default al een extra newline komt: 
<img width="361" height="169" alt="image" src="https://github.com/user-attachments/assets/f5e1f200-d42c-4059-ad28-5cbc2d4292fd" />

TimvdLippe commented:
Besproken om de volgende velden als verplicht voor te stellen:
1. status
2. title
3. detail

Optioneel zijn de velden
1. type
2. instance

Er mogen geen andere velden gespecificeerd worden. De linter moet daarom geupdate worden dat het JSON object de verplichte velden bevat, en als er meer velden zijn enkel die optioneel zijn. Elk ander veld moet een error opleveren.

sanderke committed:
Update sections/designRules.md

Co-authored-by: Tim van der Lippe <TimvdLippe@users.noreply.github.com>

sanderke committed:
problem schema members

sanderke committed:
test cases

sanderke committed:
conform feedback

sanderke committed:
Update linter/testcases/error-type-illegal-field/openapi.json

Co-authored-by: Tim van der Lippe <TimvdLippe@users.noreply.github.com>

sanderke committed:
updated expected outputs

sanderke committed:
fixed `nlgov:problem-schema-members`

joostfarla commented on sections/designRules.md:
```diff
@@ -491,6 +491,37 @@ An API is as good as the accompanying documentation. The documentation has to be
+      <dt>Statement</dt>
+      <dd>
+         <p>Error responses with HTTP status codes <code>4xx</code> or <code>5xx</code> MUST use either <code>application/problem+json</code> or <code>application/problem+xml</code> as the <code>Content-Type</code> header, and the response body MUST conform to the structure defined in [[rfc9457]].</p>
+         <p>The following fields MUST be present: <code>status</code>, <code>title</code>, and <code>detail</code>. Additionally, only these fields MAY be present: <code>type</code> and <code>instance</code>.</p>
```
Er moet ook ruimte zijn voor additionele details, bijv validatie-feedback van de request body waarbij je vaak per property een validatie-boodschap wil teruggeven. Dit zouden we ook kunnen standaardiseren of evt additionele properties toestaan.

joostfarla commented on sections/designRules.md:
```diff
@@ -491,6 +491,37 @@ An API is as good as the accompanying documentation. The documentation has to be
+      <dt>Statement</dt>
+      <dd>
+         <p>Error responses with HTTP status codes <code>4xx</code> or <code>5xx</code> MUST use either <code>application/problem+json</code> or <code>application/problem+xml</code> as the <code>Content-Type</code> header, and the response body MUST conform to the structure defined in [[rfc9457]].</p>
+         <p>The following fields MUST be present: <code>status</code>, <code>title</code>, and <code>detail</code>. Additionally, only these fields MAY be present: <code>type</code> and <code>instance</code>.</p>
```
`instance` gaat niet over de resource zelf, maar over de "occurance" van de error (ofwel een resource die meer details geeft over de error).

sanderke replied:
Voorbeeld is gewijzigd in e71d8c616724897bae682469c761f8b4a876dd61

sanderke committed:
Correctie voorbeeld "instance"
https://github.com/Logius-standaarden/API-Design-Rules/pull/251#discussion_r2315391262

TimvdLippe committed:
Verwerk feedback uit werksessie

joostfarla committed:
Error handling (#275)

* Cleanup

* Update sections/designRules.md

Co-authored-by: Tim van der Lippe <TimvdLippe@users.noreply.github.com>

* Update description of 'code' property in design rules

* Clarify error object requirements in documentation

* Documentation of `code` values

* Update wording for 'code' URI dereferencing

---------

Co-authored-by: Tim van der Lippe <TimvdLippe@users.noreply.github.com>

TimvdLippe commented:
TBD vanuit #275 van @joostfarla :

- Index vs multi-valued parameters vs collection styles
- What is required / optional?
- Can request and response encodings differ? JSON input, but XML output? (hopefully not)
- Should API's evaluate the full request body, or may it stop after the first violation?

TimvdLippe committed:
Update designRules.md

TimvdLippe committed:
Voeg Spectral linter regels toe

TimvdLippe commented:
Met de laatste commit hebben we de linter regels. De statistieken:

```
Statistics for rule nlgov:use-problem-schema:

Analyzed: 160
Passing: 47
Percentage: 29%
```

```
Statistics for rule nlgov:problem-schema-members:

Analyzed: 160
Passing: 157
Percentage: 98%
```

```
Statistics for rule nlgov:problem-schema-members-bad-request:

Analyzed: 160
Passing: 66
Percentage: 41%
```

TimvdLippe committed:
Voeg linter regel toe voor Bad Request vereiste

TimvdLippe commented:
En de regel voor het vereisen van 400 bij parameters:

```
Statistics for rule nlgov:problem-invalid-input:

Analyzed: 160
Passing: 105
Percentage: 65%
```

TimvdLippe commented:
Notulen:
- Voeg regel toe voor het uitvoeren van alle schema checks en alle errors in 1 keer terug geven
- Headers laten we buiten beschouwing, omdat die meestal in andere 4XX en 5XX errors resulteren
- Location maken we een `String` om zo te zorgen dat het zowel JSON pointer als XPath aan kan. Tevens halen we `index` weg voor query parameter errors, waar die informatie dan in de message moet. Daar moeten we dan een note voor toevoegen
- Apart hoofdstuk voor Content Negotiation, om zo iets te zeggen hoe `application/json` en `application/problem+json` te zetten. Dit vereist meer discussie en is een algemeen concern, dus pakken we apart op

TimvdLippe commented on sections/designRules.md:
```diff
@@ -491,6 +491,37 @@ An API is as good as the accompanying documentation. The documentation has to be
+   <dl>
+      <dt>Statement</dt>
+      <dd>
+         <p>Error responses with HTTP status codes <code>4xx</code> or <code>5xx</code> MUST use either <code>application/problem+json</code> or <code>application/problem+xml</code> as the <code>Content-Type</code> header, and the response body MUST conform to the structure defined in [[rfc9457]].</p>
```
Besloten om dit als een apart hoofdstuk te beschouwen.

TimvdLippe committed:
Verwerken feedback van werksessie

TimvdLippe committed:
Voeg regel toe voor het valideren van request payload met schema

TimvdLippe committed:
Enforceert 400 voor alle endpoints zonder parameters

TimvdLippe commented:
De "vereis 400 voor invalid input" check heb ik nu gefixt zodat het bij GET requests enkel als er parameters zijn, en voor alle andere endpoints. Hier nog wat extra statistieken:

Alle endpoints behalve GET:
```
Statistics for rule nlgov:problem-invalid-input:

Analyzed: 160
Passing: 155
Percentage: 96%
```

Alle endpoints en GET als het parameters heeft:
```
Statistics for rule nlgov:problem-invalid-input:

Analyzed: 160
Passing: 104
Percentage: 65%
```

TimvdLippe commented:
Deze wijziging is goedgekeurd bij het TO 2025-12-02 voor consultatie bij Kennisplatform API's. Dat houdt in dat we het hoofdstuk "Error handling" als geheel consulteren.

Omdat de slagingspercentages behoorlijk verschillen voor de verschillende regels, is het voorstel om MUST toe te passen voor >80% en RECOMMENDED voor alle anderen. Met uitzondering van de MIME type, waar er verwachte minieme impact is op implementaties van clients. Deze verwachting moet worden getoetst bij Kennisplatform API's.

TimvdLippe committed:
Fix voorbeeld

Het veld "detail" miste

TimvdLippe committed:
Merge branch 'develop' into rfc9457

TimvdLippe committed:
DELETE hetzelfde behandelen als GET voor 400 bad request

Co-authored-by: Dimitri van Hees <dimitri@vanhees.tech>

TimvdLippe commented:
Deze wijziging is gedeeltelijk goedgekeurd op het TO API van 2026-03-03. Alle regels behalve `/core/error-handling/bad-request` zijn goedgekeurd. Voor deze laatste regel wordt de discussie verder gevoerd met inachtneming van de input van Kadaster over hun huidige omgang van error handling (CC @strijm). Hier zal er expliciet worden gekeken naar de huidige oplossing bij Haal Centraal en worden gekeken naar:

1. Het mogelijk hernoemen van `errors` naar `invalid-params`
2. Analyseren van impact van `name` vervangen door `location`
3. Kijken hoe de gedetailleerde specificatie van error handling in `.feature` bestanden bij Kadaster te vatten is in de structuur

joostfarla commented:
> Het mogelijk hernoemen van errors naar invalid-params

Dit lijkt me niet wenselijk, omdat je een JSON property niet kunt beschouwen als een "param" (in tegenstelling tot query en header parameters). Hier zit dus een semantische mismatch. Om deze reden was gekozen voor de algemenere term `errors`, in lijn met de [voorbeelden in de RFC](https://www.rfc-editor.org/rfc/rfc9457.html#section-3-7) en bijv ook in [dit blog](https://swagger.io/blog/problem-details-rfc9457-api-error-handling/).

> Analyseren van impact van name vervangen door location

Het gebruik van `name` impliceert een enkelvoudige naam van een top-level property (in het geval van een `body` value). Dat lijkt ook de manier waarop het in de genoemde API's wordt gebruikt. Je wil echter ook errors kunnen rapporteren voor dieper gelegen waardes (zoals nested object properties of array items). Vandaar dat er gekozen is voor de bredere term `location` die voor `body` values een locator van de value verwacht (JSON Pointer voor JSON, XPath voor XML). Voor parameters is de waarde van `location` enkel de naam.

TimvdLippe committed:
Merge branch 'develop' into rfc9457

TimvdLippe committed:
Verwijder bad request errors regel

Deze is nog niet goedgekeurd in het TO.

TimvdLippe commented:
Ik heb #310 geopend om de discussie verder te voeren over de regel
